// @covalent/core/esm2015/common/covalent-core-common.js
/**
 * @fileoverview added by tsickle
 * Generated from: covalent-core-common.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * Generated bundle index. Do not edit.
 */
export { CovalentCommonModule, tdRotateAnimation, tdCollapseAnimation, tdFadeInOutAnimation, tdBounceAnimation, tdFlashAnimation, tdHeadshakeAnimation, tdJelloAnimation, tdPulseAnimation, mixinControlValueAccessor, mixinDisabled, mixinDisableRipple, TdAutoTrimDirective, TdFullscreenDirective, CovalentValidators, copyToClipboard, convertObjectsToCSV, convertCSVToJSON, formatJSON, downloadCSV, downloadJSON, downloadObjectsToCSV, downloadObjectsToJSON, downloadFile, readFile, TdTimeAgoPipe, TdTimeDifferencePipe, TdBytesPipe, TdDigitsPipe, TdTruncatePipe, TdDecimalBytesPipe, TdTimeUntilPipe } from './index';
export { IconService as ɵb } from './services/icon.service';
export { RouterPathService as ɵa } from './services/router-path.service';
