// @covalent/core/esm2015/common/directives/fullscreen/fullscreen.directive.js
/**
 * @fileoverview added by tsickle
 * Generated from: directives/fullscreen/fullscreen.directive.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, HostListener, ElementRef, Inject } from '@angular/core';
import { DOCUMENT } from '@angular/common';
/**
 * @record
 */
function IFsDocument() { }
if (false) {
    /** @type {?} */
    IFsDocument.prototype.fullscreenElement;
    /** @type {?} */
    IFsDocument.prototype.webkitFullscreenElement;
    /** @type {?} */
    IFsDocument.prototype.mozFullscreenElement;
    /** @type {?} */
    IFsDocument.prototype.msFullscreenElement;
    /** @type {?} */
    IFsDocument.prototype.webkitExitFullscreen;
    /** @type {?} */
    IFsDocument.prototype.mozCancelFullScreen;
    /** @type {?} */
    IFsDocument.prototype.msExitFullscreen;
}
export class TdFullscreenDirective {
    /**
     * @param {?} _document
     * @param {?} _el
     */
    constructor(_document, _el) {
        this._document = _document;
        this._el = _el;
        this.fullScreenIsActive = false;
    }
    /**
     * @param {?} event
     * @return {?}
     */
    fsChangeHandler(event) {
        this.fullScreenIsActive = event.srcElement === this._getFullScreenElement();
    }
    /**
     * @return {?}
     */
    toggleFullScreen() {
        this._getFullScreenElement() === this._el.nativeElement ? this.exitFullScreen() : this.enterFullScreen();
    }
    /**
     * @return {?}
     */
    enterFullScreen() {
        const { _el: { nativeElement }, } = this;
        /** @type {?} */
        const enterFullScreenMap = {
            requestFullscreen: (/**
             * @return {?}
             */
            () => nativeElement.requestFullscreen()),
            // Chrome
            webkitRequestFullscreen: (/**
             * @return {?}
             */
            () => nativeElement.webkitRequestFullscreen()),
            // Safari
            mozRequestFullScreen: (/**
             * @return {?}
             */
            () => nativeElement.mozRequestFullScreen()),
            // Firefox
            msRequestFullscreen: (/**
             * @return {?}
             */
            () => nativeElement.msRequestFullscreen()),
        };
        for (const handler of Object.keys(enterFullScreenMap)) {
            if (nativeElement[handler]) {
                enterFullScreenMap[handler]();
            }
        }
    }
    /**
     * @return {?}
     */
    exitFullScreen() {
        const { _document, _el: { nativeElement }, } = this;
        /** @type {?} */
        const exitFullScreenMap = {
            exitFullscreen: (/**
             * @return {?}
             */
            () => _document.exitFullscreen()),
            // Chrome
            webkitExitFullscreen: (/**
             * @return {?}
             */
            () => _document.webkitExitFullscreen()),
            // Safari
            mozCancelFullScreen: (/**
             * @return {?}
             */
            () => _document.mozCancelFullScreen()),
            // Firefox
            msExitFullscreen: (/**
             * @return {?}
             */
            () => _document.msExitFullscreen()),
        };
        for (const handler of Object.keys(exitFullScreenMap)) {
            if (_document[handler] && this._getFullScreenElement() === nativeElement) {
                exitFullScreenMap[handler]();
            }
        }
    }
    /**
     * @private
     * @return {?}
     */
    _getFullScreenElement() {
        const { _document } = this;
        /** @type {?} */
        const tdFullScreenElementMap = {
            fullscreenElement: (/**
             * @return {?}
             */
            () => _document.fullscreenElement),
            // Chrome, Opera
            webkitFullscreenElement: (/**
             * @return {?}
             */
            () => _document.webkitFullscreenElement),
            // Safari
            mozFullscreenElement: (/**
             * @return {?}
             */
            () => _document.mozFullscreenElement),
            // Firefox
            msFullscreenElement: (/**
             * @return {?}
             */
            () => _document.msFullscreenElement),
        };
        for (const props of Object.keys(tdFullScreenElementMap)) {
            if (_document[props]) {
                return _document[props];
            }
        }
    }
}
TdFullscreenDirective.decorators = [
    { type: Directive, args: [{
                selector: '[tdFullScreen]',
                exportAs: 'tdFullScreen',
            },] }
];
/** @nocollapse */
TdFullscreenDirective.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] },
    { type: ElementRef }
];
TdFullscreenDirective.propDecorators = {
    fsChangeHandler: [{ type: HostListener, args: ['document:fullscreenchange', ['$event'],] }, { type: HostListener, args: ['document:webkitfullscreenchange', ['$event'],] }, { type: HostListener, args: ['document:mozfullscreenchange', ['$event'],] }, { type: HostListener, args: ['document:msfullscreenchange', ['$event'],] }]
};
if (false) {
    /** @type {?} */
    TdFullscreenDirective.prototype.fullScreenIsActive;
    /**
     * @type {?}
     * @private
     */
    TdFullscreenDirective.prototype._document;
    /**
     * @type {?}
     * @private
     */
    TdFullscreenDirective.prototype._el;
}
